[28기 안현재] Mission 완료 후 리뷰 반영중

hyeonze opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [x] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 인스타그램의 로그인 및 메인페이지 레이아웃과 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 입력 데이터 유효성 검사 && 검사에 따른 로그인버튼 활성화
- 입력창에 변화가 있을 때마다 idInput/pwInput에 저장
- useEffect에서 idInput/pwInput에 변동이 있을 경우 유효성검사를 해서 결과를 불리언타입으로 isValidatedId/isValidatedPw에 저장
- useEffect에서  isValidatedId/isValidatedPw에 변동이 있을때마다 activateBtn를 ture/false로 변경
- activateBtn이 ture면 로그인버튼에 className이 추가되고 disabled가 false가 됨(버튼 활성화)
- activateBtn이 false면 로그인버튼에 className이 빈 문자열이 되고 disabled가 true가 됨(버튼 비활성화)
- 테스트 방법 : `/login` 페이지 이동 >>> 로그인 input 데이터 입력

2. Main
- 좌측 Feeds 컴포넌트, 우측 MainRight컴포넌트로 분리
- Feeds는 map 함수를 통해 feedData.json으로부터 받아온 배열의 길이만큼 리턴
 
3. Feeds
- 좌측 피드를 담당하는 컴포넌트로, 댓글을 관리하는 자식컴포넌트 Comments가 있음
- Main에서 부여한 props(key, userId, img)전달받아 사용
- 댓글에 내용이 들어오면 commentVal에 저장
- commentVal에 내용이 있으면 classOfBtn을 'blue'로 변경해 게시버튼 활성화
- 댓글창에서 enter을 입력하거나 게시버튼을 누르면 uploadComment함수를 통해 댓글내용 추가
- 추가 후 댓글창 초기화 및 게시버튼 비활성화

4. Comments
- 댓글을 map을 통해 반복구현하는 컴포넌트
- 부모 컴포넌트인 Feeds로부터 댓글 내용이 담긴 배열 commentVal을 props로 받아 사용
- commentVal 내부의 id(key), userId, value, time을 사용해 댓글 구현
- 댓글 내용에서 직접 map메서드를 사용하는 것을 방지하기 위해 하위 컴포넌트로 Comment추가

5. MainRight
- Main의 우측 사이드바를 담당하는 컴포넌트
- styles/_color.scss에서 정의한 팀원 공통 그라데이션 적용

6. Mock data
- public > data 폴더에 commentData.json/feedData.json 파일을 저장해 목업 데이터를 불러왔을 때 화면을 테스트 가능

7. 댓글 좋아요 및 삭제기능
- 각 댓글의 좋아요/삭제 버튼을 클릭 시 toggleLike/handleDelete함수 실행
- toggleLike함수는 isLiked라는 state를 true/false로 토글해주며 좋아요를 제어함
- handleDelete함수는 feeds를 통해 props로 받은 함수고, 내부에서 filter함수를 통해 댓글을 제거
<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- useState를 사용해 변수를 선언하고 변수를 조작하고 싶을 때 사용할 수 있게 됐습니다.
- useEffect를 사용해 주시할 변수를 의존성 배열에 지정하고, 해당 변수에 변동이 있을 경우 발생할 실행문을 추가할 수 있게 됐습니다.
- Mock data를 활용해 백엔드와 협업 전에 프로젝트가 어떻게 실행될 지 예측해보고, 시험해볼 수 있게 됐습니다.
- 댓글 좋아요 및 삭제기능을 추가하면서 컴포넌트 상하위 구조를 바꿔야할 때 조금 더 침착하게 대응할 수 있게 됐습니다.

<br />

## :: 기타 질문 및 특이 사항
- map으로 컴포넌트 반복구현 시 궁금한점 있습니다.(질의 완료)
map 사용해서 컴포넌트 반복 구현 시 부모컴포넌트에서 해도 되고, 자식컴포넌트에서 해도 되는데 차이점이 있는지. 이 때 map을 실행하는 과정에서 key값을 주게 되는데, 부모컴포넌트에서 key값을 주는 것과 자식컴포넌트에서 key값을 주는것이 차이가 있는지.(React Developer Tools 상에서는 차이가 있어 보이는데 유의미한 차이일지. 현재 프로젝트에서는 다중Feed구현 시 전자의 방법을 사용했고, 댓글 구현 시 후자의 방법 사용했음.) (답변)=> 코드 가독성(추상화?)에서 차이가 있고 차후 Suspense를 고려하면 따져줘야 할 수 있으나, 현재 레벨에서는 같은 코드이므로 둘 다 사용해도 좋음.

Commit:
First commmit

Commit:
Merge branch 'master' into feature/hyeonze

Commit:
Add: Mission 2 - 로그인버튼 활성화 기능 구현

Commit:
Fix : Mission 2 로직 삼항연산자로 수정

Commit:
Backup : 댓글 컴포넌트화 하기 전 백업

Commit:
Backup : 컴포넌트 분리 후 백업

Commit:
Add : Mission 4 ) 댓글 컴포넌트화 + props로 데이터 전달

Commit:
[Add] : Mission 3 - 댓글 기능 구현
해결못한점 : 비동기 대응(댓글 추가 후 바로 업로드 되지 않음), 댓글 추가 후 input.value초기화 안됨

Commit:
Backup : 컴포넌트 분리 전 백업

Commit:
Backup : <main>제거 후 bkup

Commit:
Backup : #wrap 제거 후 bkup

Commit:
Fix : 태그 정리 후 반응형 대응

Commit:
Fix : 로그인버튼 활성화기능 삼항연산자 사용한 문법으로 수정

Commit:
Fix : 필요없는 usestate 사용 제거(addedCommentVal)후 변수로 변경

Commit:
Backup : 로그인페이지 로직변경 전 백업

Commit:
Fix : 로그인버튼 비동기로 늦게 활성화되는 에러  처리

Commit:
Add : 댓글 입력 후 내용 초기화 및 게시버튼 비활성화 기능 추가

Commit:
Fix : feeds 다중배치 전 레이아웃 수정

Commit:
[Add & Fix] Mission2 다중 Feed를 fetch로 받기 구현 && 피드 우측상단 트리플닷 아이콘 안보이는 현상 수정
mock data를 활용한 여러개의 피드 구현 완료
피드 우측상단 트리플닷 아이콘 안보이는 현상 수정

Commit:
Fix : common.scss 잠시 주석 처리했던 것 해제 후 레이아웃 수정

Commit:
Fix : 로그인페이지 정보 받아올 때 비구조화 할당 적용, 로그인페이지 레이아웃 수정

Commit:
Fix : 피어리뷰 전 필요없는 주석 제거

Commit:
Fix : 빈 댓글창에서 엔터 입력 시 빈 댓글 입력되는 버그 수정

Commit:
Fix : 필요없는 event인자 제거

Commit:
Fix : 안쓴 el 인자 사용해서 코드 가독성 향상

Commit:
Fix : 안쓴 el 인자 사용해서 코드 가독성 향상

Commit:
Add : 피드별 사진 추가

YOONKY commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,76 @@
+  };
+
+  useEffect(() => {
+    idInput.indexOf('@') !== -1
```
이 두 조건을 &&로 걸면 state를 하나만 사용하고 validation을 확인할 수 있지 않을까요? 
왜 유즈이펙트를 쓰셨는지 궁금합니다.

hyeonze replied:
이 두 조건을 &&로 걸면 state를 하나만 사용하고 validation을 확인할 수 있지 않을까요? => 맞는 것 같습니다.

수정 전 코드에서 유효성검사를 changeIdInput, changePwInput 내에서 각각 해줬었는데 호환성 검사에 딜레이가 생겨서 useEffect를 사용하는게 해결책이라 생각했습니다. 이 과정에서 changeIdInput, changePwInput 내부의 삼항연산자들을 useEffect 내부로 잘라내어 오게 되면서 로직이 길어진 것 같습니다.

YOONKY commented on src/pages/hyeonze/Main/MainRight/MainRight.scss:
```diff
@@ -0,0 +1,193 @@
+      padding: 10px;
+    }
+
+    > span:first-child {
```
'>'이걸 사용하신 특별한 이유가 있으신가요?

hyeonze replied:
우측 추천파트에서 <span>회원님을 위한 추천</span> 에 적용되는 코드인데, 자식선택자를 제거하게 되면 .recommended 하위의 모든 첫 번째 자식인 span에(user5, user6, user7) 모두 적용됩니다. 더 좋은 방법 있으면 추천해주세요!

YOONKY reviewed (commented):
고생하셨습니다! 여러 피드까지 구현하신 점이 대단해요!👍

Commit:
Fix : 댓글내용 한개의 json파일에서 피드별로 관리할 수 있도록 개선

joonyg10 commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,76 @@
+  const [activateBtn, setActivateBtn] = useState(false);
+  const [isValidatedId, setIsValidatedId] = useState(false);
+  const [isValidatedPw, setIsValidatedPw] = useState(false);
+
```
id와 pw로 한번에 validation하고 activateBtn에 True/false를 지정하면 될 것 같습니다. isValidatedId나 isValidatedPw는 크게 필요하지 않을 것 같아요. 불필요한 리랜더링도 줄일 수 있다고 생각하는데 어떻게 생각하시나요?

Commit:
Fix : 목업데이터 내용 구체화

joonyg10 commented on src/pages/hyeonze/Login/Login.scss:
```diff
@@ -0,0 +1,68 @@
+          border-radius: 3px;
+          font-size: 16px;
+        }
+
```
input이 inline-block이고 수직으로 정렬하려고 display: flex를 쓰신 것 같습니다. 굉장히 좋은데요 저는 Input을 display: block해주는 것도 괜찮은 방법이지 않을까 생각합니다.

joonyg10 commented on src/pages/hyeonze/Login/Login.scss:
```diff
@@ -0,0 +1,68 @@
+          height: 60px;
+          margin-top: 16px;
+          align-items: center;
+          justify-content: center;
```
'로그인' 이라는 글자 센터에 맞추기 위해 display: flex주신 것 같아요. 좋은방법이지만 flex와 그 속성들이 좀 떨어져 있는 것 같습니다. 이렇게 연관되는 css속성은 붙여 주시는 게 나중에 유지보수 측면에서도 좋을 것 같아 추천 드립니다!

Commit:
Fix : 미사용 데이터 삭제

Commit:
Fix : eslint워닝 및 주석 삭제

joonyg10 commented on src/pages/hyeonze/Main/Feeds/Comment/Comment.jsx:
```diff
@@ -0,0 +1,33 @@
+
+  return (
+    <>
+      {commentVal.map((el, i) => {
```
저번 세션에서 멘토님이 하신 것처럼 변수명을 조금 더 구체적으로 어떤 내용을 담고 있는지를 작성해 주시면 좋을 것 같습니다. commentVal이 댓글 전체를 관리하는 배열이고 el이 각 배열의 요소이니깐 
commentVal대신 allComments, el대신 comment등으로 (이건 저도 조금 애매하네요 .. 죄송합니다) (배열은 s를 붙인 변수명으로 하는 게 좋을 것 같습니다!)

joonyg10 commented on src/pages/hyeonze/Main/Feeds/Comment/Comment.scss:
```diff
@@ -0,0 +1,48 @@
+span {
```
span에 대해서만 스타일링을 해주셨는데 이러면 다른 컴포넌트의 span에도 영향이 가지 않을까 조심스럽게 생각해 봅니다. 개인적으로 className을 주는 게 조금 더 안전할 것 같습니다

 추가적으로 svg태그도 span 내부에 nesting해주시면 좋을 것 같아요.
```
svg {
   &.red {} 
   &.blue {} 
} 
```

joonyg10 commented on src/pages/hyeonze/Main/Feeds/Feeds.jsx:
```diff
@@ -0,0 +1,91 @@
+    setCommentVal(addedCommentVal);
+    setCurrInputVal('');
+    setClassOfBtn('');
+  };
```
굉장히 잘 짜신 것 같습니다!! 👏 그런데 순서가 조금 애매한 것 같아요. 17)에서 uploadComment 호출을 하고 31)에서 currInputVal( " " ) -> 18 ) 에서 currInputVal( e.target.value ) 순으로 진행이 된다면 댓글을 추가하더라도 input은 비워지지 않을 것 같습니다.

joonyg10 commented on src/pages/hyeonze/Main/Feeds/Feeds.scss:
```diff
@@ -0,0 +1,107 @@
+    }
+  }
+
+  article {
```
Feeds.jsx에서 .feeds의 바로 자식요소가 article이라 article을 .feeds 바로 다음에 써주는 게 좋을 것 같습니다. 그리고 height요소는 자식요소들의 높이에 따라 결정되니깐 height: 100%를 주어도 feeds의 높이가 변하지는 않을 것 같다는 생각이 드는데 혹시 설정하신 이유가 있으신가요?

joonyg10 commented on src/pages/hyeonze/Main/MainRight/MainRight.scss:
```diff
@@ -0,0 +1,193 @@
+    @include mixin.flex($justify: space-between, $align: flex-start);
+    overflow: hidden;
+    margin-top: 20px;
+    width: 100%;
```
.stories자체가 div여서 width:100% 안주어도 될 것 같은데 혹시 너비가 고정이 안되어서 지정하신 건가요?

Commit:
Bkup : 로그인/회원가입 실습 전 백업

Commit:
Bkup : 로그인/회원가입 실습 후 백업(라이브리뷰준비)

Commit:
Fix : 댓글 좋아요 기능 수정
댓글 좋아요 누르면 전체 하트에 좋아요가 들어가는 버그 수정했습니다

bonbon0202 commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,115 @@
+  const [pwInput, setPwInput] = useState('');
+  const [activateBtn, setActivateBtn] = useState(false);
+  const [isValidatedId, setIsValidatedId] = useState(false);
+  const [isValidatedPw, setIsValidatedPw] = useState(false);
```
[공식문서 React로 사고하기](https://ko.reactjs.org/docs/thinking-in-react.html#step-3-identify-the-minimal-but-complete-representation-of-ui-state) 를 보시면 어떤 값들이 state가 되어야하는지에 대해 적혀있습니다. 각각 살펴보고 해당 데이터는 state로 적절한지 댓글로 남겨주세요. 각 데이터에 대해 아래의 세 가지 질문을 통해 결정할 수 있습니다

> 1. 부모로부터 props를 통해 전달됩니까? 그러면 확실히 state가 아닙니다.
> 

> 2. 시간이 지나도 변하지 않나요? 그러면 확실히 state가 아닙니다.
> 

> 3. 컴포넌트 안의 다른 state나 props를 가지고 계산 가능한가요? 그렇다면 state가 아닙니다.
>

bonbon0202 commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,115 @@
+    isValidatedId && isValidatedPw
+      ? setActivateBtn(true)
+      : setActivateBtn(false);
+  }, [idInput, pwInput, isValidatedId, isValidatedPw]);
```
라이브리뷰 때 말씀드린 것처럼 조건의 falsy, truthy한 값을 이용해 리팩토링 해주세요! :) 
또는 state를 줄일 경우 불필요할 수도 있겠네요

bonbon0202 commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,115 @@
+  // };
+
+  return (
+    <div className="login">
```
.

bonbon0202 commented on src/pages/hyeonze/Login/Login.jsx:
```diff
@@ -1,5 +1,115 @@
+            onChange={changePwInput}
+          />
+          {/* <Link to="/main">로그인</Link> */}
+          <input
+            className={activateBtn ? 'blue' : ''}
+            disabled={activateBtn ? false : true}
+            type="button"
+            value="로그인"
+            onClick={goToMain}
+            // onClick={signup}
+            // onClick={login}
+          />
```
.

bonbon0202 replied:
- 불필요한 주석처리는 삭제해주세요!
- 더불어서 className 자체를 '색'으로 지정하는 것은 좋은 네이밍은 아닌 것 같습니다. 어떤 네이밍이 적합할지 고민하셔서 , 해당 요소가 무엇인지를 잘 나타내는 네이밍으로 수정해주세요. 
- 플러스! 여기에도 사용하지 않아도 되는 삼항연사자가 숨어있네요! 확인해주세요

bonbon0202 commented on src/pages/hyeonze/Login/Login.scss:
```diff
@@ -0,0 +1,68 @@
+@use '../../../styles/color.scss' as clr;
+
+body {
+  background: clr.$bg-login;
```
위 코드는 어디에 작성되면 좋을까요?

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Comment/Comment.jsx:
```diff
@@ -0,0 +1,26 @@
+  };
+
+  return (
+    <span>
```
.

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Comment/Comment.scss:
```diff
@@ -0,0 +1,48 @@
+  }
+}
+svg.red {
+  color: #f00;
```
자주 사용이 되는 컬러값이네요! 이런 경우 sass variables 기능을 이용하는것은 어떨까요?

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Comment/Comments.jsx:
```diff
@@ -0,0 +1,11 @@
+      <Comment key={el.id} userId={el.userId} value={el.value} time={el.time} />
+    );
+  });
+}
```
.

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Feeds.jsx:
```diff
@@ -0,0 +1,81 @@
+  const [commentVal, setCommentVal] = useState(comments);
+  const addedCommentVal = [...commentVal];
+  const [classOfBtn, setClassOfBtn] = useState('');
+  const [currInputVal, setCurrInputVal] = useState('');
```
.

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Feeds.jsx:
```diff
@@ -0,0 +1,81 @@
+    value ? setClassOfBtn('blue') : setClassOfBtn('');
+    if (e.keyCode === 13) uploadComment();
+    setCurrInputVal(value);
+  };
```
.해당 부분도 좀 더 반복되는 단어를 줄일 수 없을지 고민해봅시다. 
더불어 불필요한 state값도 보이네요! 위 리뷰를 참고해 수정해주세요!

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Feeds.scss:
```diff
@@ -0,0 +1,107 @@
+    }
+  }
+
+  .add_comments {
```
class명은 어떤 형태가 좋을까요?! 
해당 class명은 '동작'을 나타내기에 클래스 네임보다는 함수명처럼 느껴집니다!

bonbon0202 commented on src/pages/hyeonze/Main/Main.jsx:
```diff
@@ -1,5 +1,39 @@
+        for (let i in data) currInitialComment.push(data[i].comment);
+      })
+    );
+  });
```
.

bonbon0202 commented on src/pages/hyeonze/Main/MainRight/MainRight.jsx:
```diff
@@ -0,0 +1,89 @@
+        <span>스토리</span>
+        <span>모두보기</span>
+        <ul>
+          <li>
+            <img src="images/hyeonze/user1.jpg" alt="" />
+            <div>
+              <span>user1</span>
+              <span>1시간전</span>
+            </div>
+          </li>
+          <li>
+            <img src="images/hyeonze/user2.jpg" alt="" />
+            <div>
+              <span>user2</span>
+              <span>2시간전</span>
+            </div>
+          </li>
+          <li>
+            <img src="images/hyeonze/user3.jpg" alt="" />
+            <div>
+              <span>user3</span>
+              <span>3시간전</span>
+            </div>
+          </li>
+          <li>
+            <img src="images/hyeonze/user4.jpg" alt="" />
+            <div>
+              <span>user4</span>
+              <span>4시간전</span>
+            </div>
+          </li>
```
반복되는 UI들이 보이네요! 반복되는 UI를 어떻게 줄일 수 있을까요?

Commit:
Add : 댓글 삭제기능 추가

Commit:
Fix : 댓글 삭제기능 에러 수정

Commit:
Fix : handleDelete 함수 수정 통한 가독성 향상

bonbon0202 commented on src/pages/hyeonze/Main/Feeds/Feeds.scss:
```diff
@@ -0,0 +1,107 @@
+  top: 77px;
+  width: 60%;
+  margin-bottom: 80px;
+  background: #fff;
+
+  .top_menu {
+    @include mixin.flex($justify: space-between, $align: center);
+    height: 82px;
+    background: #fff;
```
이렇게 각 요소에 background color를 각각 다 넣어주신 이유가 있으실까요?

Commit:
Fix : 주석 제거 및 훅 위치 변경

Commit:
Fix : 로그인 페이지 불필요한 state 제거 및 클래스이름 변경 및 주석제거 및 코드순서 변경
리뷰반영하여 state 1개로 로그인 버튼 활성화기능 제어
리뷰반영하여 클래스이름 blue >> activated 수정
필요없는 주석 제거
Hooks 하단에 모아서 작성

Commit:
Fix : 필요없는 useEffect 의존성배열 요소 제거

Commit:
Fix : useEffect로직 단순화

Commit:
Fix : isValidatedUser && 'activated'로 했을 경우 Warning 떠서 삼항연산자로 수정

Commit:
Fix : Login.scss form 가운데정렬 방법 수정

Commit:
Fix : Login.scss flex관련속성 순서 수정

Commit:
Fix : 리뷰반영하여 클래스이름 및 네스팅 수정

Commit:
Fix : comment 게시버튼 활성화 로직 수정 및 클래스이름 수정

Commit:
Bkup : Feeds.scss 수정 전 백업(고칠것 많음)
